TinyMCE Comments Plugin

domwhewell-sage opened the pull request:
Draft PR for a comments plugin mentioned in #425 
Added a new tinymce plugin to the folder called `ghostwriter-comments`
- The plugin adds a `Comment` context menu item to all wysiwyg editors
<img width="453" alt="image" src="https://github.com/GhostManager/Ghostwriter/assets/122788350/d23327e0-c3a2-4eea-b41b-486705c23cde">

- It'll allow you to highlight some text and add a comment. 
- Once a comment is added it'll appear with a red background until you select resolve. 
<img width="174" alt="image" src="https://github.com/GhostManager/Ghostwriter/assets/122788350/9bd3bc9d-ad03-4140-a890-4d0e767ed01e">

- Once a comment is resolved the red background disappears but the comment remains in the source code of the editor the comment thread can also be viewed again by highlighting the text and selecting comment
```html
<p>This is a <span class="comment-resolved" comment-data="%5B%7B%22author%22%3A%2261646d696e%22%2C%22comment%22%3A%22Test%22%7D%2C%7B%22author%22%3A%22646f6d2e77686577656c6c%22%2C%22comment%22%3A%22Hello%22%7D%5D">test</span> vulnerability</p>
```

- Multiple collaborators can comment on findings. The comment from the current user will always appear on the left
<img width="362" alt="image" src="https://github.com/user-attachments/assets/3387fa67-b721-4eda-bfc5-84d22dc8ac51">


The comment thread currently appears with the commenters username avatar and then their comment.
```html
<div class="${selfclass}">
    <div class="comment-header">
        <h1>${tinymce.DOM.encode(hex2ascii(author))}</h1>
        <img class="comment-avatar" src="/users/${tinymce.DOM.encode(hex2ascii(author))}/avatar" alt="Avatar">
    </div>
    <p>${tinymce.DOM.encode(comment)}</p>
</div>
```

The comments dont currently transfer over to the docx but Im not sure they should, the reportwriter is currently just adding the paragraph element into the docx correctly without the comment so we might not want to change that

domwhewell-sage commented:
I'm trying to think of a way to notify the author which findings / fields have comments on them sort of like the "Previous" / "Next" functionality in word

<img width="106" alt="image" src="https://github.com/GhostManager/Ghostwriter/assets/122788350/2a78ec07-2a79-4111-bee7-bdcd5af41a12">

We could potentially send an ajax request to set the finding as "Needs Editing" once a reviewer has posted a comment on a finding but I think sending a list of urls to the user that take them directly to each comment would be more useful. We might be able to use the websockets notifications for the user but I don't know how long these stay available to the user

ColonelThirtyTwo commented on ghostwriter/static/js/tinymce/plugins/ghostwriter-comments/plugin.min.js:
```diff
@@ -0,0 +1,106 @@
+                `<div id="comment-thread" style="float: ${float};">
+                    <div class="comment-header">
+                        <h1>${hex2ascii(author)}</h1>
+                        <img class="comment-avatar" src="/users/${hex2ascii(author)}/avatar" alt="Avatar">
```
Both this and the line above should use `${tinymce.DOM.encode(hex2ascii(author))}` to prevent HTML injection.

ColonelThirtyTwo commented on ghostwriter/static/js/tinymce/plugins/ghostwriter-comments/plugin.min.js:
```diff
@@ -0,0 +1,106 @@
+                const comment = item.comment;
+                const float = author === username ? 'right' : 'left';
+                commentsThread +=
+                `<div id="comment-thread" style="float: ${float};">
```
Since we're adding CSS to the CSS file above for avatars, we may as well provide the comment thread styles there too.

I recommend using [flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) instead of clear here. Something like this, though you'll need to wrap the whole thread in a div:
```css
#entire-comment-thread {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
}
#entire-comment-thread .comment {
    margin-left: 0;
    margin-right: 20px;
}
#entire-comment-thread .comment.from-self {
    align-self: flex-end;
    margin-left: 20px;
    margin-right: 0;
}
```

(Also as-is, this code emits multiple elements with the same ID)

domwhewell-sage replied:
Thanks, I agree using flexbox is better than clear. I have wrapped it all in the `entire-comment-thread` div and set the class of each comment depending on the author.
I am still unsure of the styling here...

domwhewell-sage replied:
I've also added a border around each individual comment and positioned the avatar on the top right of this comment border

ColonelThirtyTwo commented on ghostwriter/static/css/styles.css:
```diff
@@ -239,6 +239,15 @@ hr {
+  border-width: 1px !important;
+  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19) !important;
+}
+
```
I don't think these need to be !important.

ColonelThirtyTwo commented on ghostwriter/static/js/tinymce/plugins/ghostwriter-comments/plugin.min.js:
```diff
@@ -0,0 +1,106 @@
+        icon: 'comment',
+        text: 'Comment',
+        onAction: function () {
+            const selection = editor.selection.getContent();
```
Check if `selection` is empty here after assignment and return if not, rather than waiting until the thread HTML is generated.

domwhewell-sage replied:
Added an if statement to return if `selection` is null

ColonelThirtyTwo reviewed (changes requested):
Thank you for your patch! I've added a few notes - let me know if you're still interested in working on this.

domwhewell-sage commented:
Thanks @ColonelThirtyTwo I have made the requested changes. And added a border around each individual comment, moved the avatar to the top right of the comment.

I am still interested in working on this however I am still trying to think of a way to solve the notification problem. After your report has been QA'd (and comments have been left) it might be time consuming for the author to go back through each individual issue and check if all comments have been addressed in the report.

domwhewell-sage commented:
Thinking about it I don't think the notifications should be part of this plugin as I did have it sending websocket messages once a comment had been added but if the user didn't submit the form the comment wouldn't save therefore notifications should probably be done server side once the form is submitted.

Something like this: If the updated form contains any rich text fields with 1 or more `<span class="comment">` tags in the source send a notification.

Marking this as ready for review now.

ColonelThirtyTwo commented on ghostwriter/static/js/tinymce/plugins/ghostwriter-comments/plugin.min.js:
```diff
@@ -0,0 +1,110 @@
+    function createComment(comment, selection) {
+        const commentNode = editor.selection.getNode();
+        let commentThread = readComments(commentNode);
+        commentThread.push({ author: username, comment: comment });
```
I'm curious as to where the `username` variable is set, and also why it's in hex. We should be able to just provide it as a normal string. I'm also not sure if it will be set in the admin pages. I'll look into it.

domwhewell-sage replied:
The username is a constant set on the `base_generic.html` and `base_generic_empty.html` template so I think it should be everywhere (maybe...?)

ColonelThirtyTwo replied:
Yeah, found it.

It's not provided if a user isn't logged in, but we don't have any rich text fields for anonymous users, so whatever.

Admin pages won't use those templates, though I don't think we need to have comments anywhere in there, so we can probably just disable comments if there's no username.

Seems like the section it's defined in is setting up the websocket notifiers, and the variables it uses are just leaked to the global scope, which I'm not a fan of. But that's a different issue.

ColonelThirtyTwo commented on ghostwriter/static/js/tinymce/plugins/ghostwriter-comments/plugin.min.js:
```diff
@@ -0,0 +1,110 @@
+            if (selection === '') {
+                return;
+            }
+            const node = editor.selection.getNode();
```
This should search the selected node's ancestors for comment nodes as well - otherwise, selecting or clicking on a subspan will not work properly (ex. if you comment on a whole paragraph but click on a bolded section).

ColonelThirtyTwo commented on ghostwriter/static/js/tinymce/plugins/ghostwriter-comments/plugin.min.js:
```diff
@@ -0,0 +1,110 @@
+                const author = item.author;
+                const comment = item.comment;
+                commentThread.push({ author: author, comment: comment });
+            });
```
Shouldn't need to make a new array here - you should be able to just do `return commentData`.

ColonelThirtyTwo commented on ghostwriter/static/js/tinymce/plugins/ghostwriter-comments/plugin.min.js:
```diff
@@ -0,0 +1,110 @@
+        const commentNode = editor.selection.getNode();
+        let commentThread = readComments(commentNode);
+        commentThread.push({ author: username, comment: comment });
+        editor.execCommand('mceInsertContent', false, `<span class="comment" comment-data="${encodeURIComponent(JSON.stringify(commentThread))}">${selection}</span>`);
```
This assumes the selection contains the entire comment. This is true when clicking on the comment, since the handler will select the area, but it isn't true for commenting on a section with a resolved comment, or when using the keyboard to manually select a section of text, and will lead to nested comment spans.

ColonelThirtyTwo reviewed (commented):
If we want to support nested comments (which I think we should), I think this this may need to be split up into a "View Existing Comment" button and a "Create New Comment" button. View would traverse up the node tree and open the editor for an existing comment if found and replace that span specifically when edited, whereas Create would create a comment for the highlighted span. This would also allow one to edit the content of a commented span, which is impossible with the current design.

This would mean you could comment on an entire paragraph while also being able to comment on individual words or sentences.

Styling may be a pain though to indicate nested comments... you'd have to have `.comment { color: onecolor; }` and `.comment .comment { color: anothercolor; }` and so on.

domwhewell-sage commented:
Hi @ColonelThirtyTwo I think we should support nested comments too! I have made a change to support editing your comments in a thread. But I am stuck on how to create the nested comments.... I will not have too much time available to work on this for the next few weeks.

domwhewell-sage commented:
Hi @ColonelThirtyTwo, I have reworked the comment plugin to allow for nested comments, There is now a "Comments" submenu in the context menu, which allows you to "Create comment" and "View comments". "Create" is a simple dialog to create a brand new comment (Which can be nested (There is CSS to support up to three levels of comment nesting)) and "View" opens a dialog displaying the existing comments.
